RFC-1 (#1)
varias-io/standards: RFCs/RFC_1.md

# RFC: PHP Coding Standards

**Primary author(s)**: Giovan Angela

**Created**: 04-10-2022

**Last updated**: 04-10-2022

**Approver(s)**:

| Name | yes/not yet | date |
| ---- | ----------- | ---- |
|      | [ ]         |      |

## Overview
<!-- If someone only reads this far, what do you want them to know? -->

We should use some company wide standards in order to keep the styling of code as clean and consistant as possible per project, developer and environment. Therefore this RFC proposes to start using [PSR-12](https://www.php-fig.org/psr/psr-12/) as the standard for the PHP codebase.

## Goal
<!-- What problems are you trying to solve? What problems are you not trying to solve? -->

The aim of this proposal is to ensure every project and developer conforms to the same styling specification. This will help lower the amount of file changes that can be made by developers when they use their custom code formatter. This will also declutter the git history when used correctly.

> 

Reviewer: > The aim of this proposal is to ensure every project and developer conforms to the same styling specification.

I personally don't like this. I would like to suggest to split this in two parts:

- to  aim that developers in the same project conform to the same code style specfications
- to provide new projects with a extensible starting point for standards

## Motivation
<!-- What is the current state of the world? Why is this change being proposed? -->

> There is currently no standard code style being used, outside of the Summa project. Developers use different code styles that are common by their respective IDE and this may lead to repeated file changes that are purely style changes. 

Reviewer: ```suggestion
There are currently not a lot of coding style being used. Developers use different code styles that are common by their respective IDE and this may lead to repeated file changes that are purely style changes. 
```

There is more going on in other projects.

## Design
[PHP-CS-Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer) will be used to check and fix the style of the php code that has been commited.

> Using a github workflow action, found in the `.github` folder, we'll enable the check after each commit on a PR. This workflow will check the files that have been changed compared to the main branch.

Reviewer 2: In case your PR is targeting a branch other than `main` ;)

```suggestion
Using a github workflow action, found in the `.github` folder, we'll enable the check after each commit on a PR. This workflow will check the files that have been changed compared to the target branch.
```

Also added is a composer command, found in `composer.json`, to check and fix the code styling on a local machine using `composer phpcs` and `composer phpcs-fix` respectively.

## Drawbacks
<!-- What are the possible drawbacks of this proposal? ex. Higher costs, less flexibility for the developer, etc... -->
The drawbacks for this proposal could be:

For existing projects:

> - Inconvenient to set up correctly

Reviewer 3: Why?

Author: Configuration is a bitch sometime

> - One PR with an enormous amount of file changes related to styling

Reviewer 3: This is not the case if you only apply it to the changes right?

Reviewer: You can also ignore this in git history if you want, so I don't really see the cons here.
